File: backend/src/captionovo/services/processing_events.py
from datetime import UTC, datetime


class ProcessingEventsService:
    def __init__(self, admin) -> None:
        self._admin = admin

    def emit(
        self,
        *,
        project_id: str,
        stage: str,
        status: str,
        job_id: str | None = None,
        message: str | None = None,
        metadata: dict | None = None,
    ) -> None:
        completed_at = None
        if status in ("completed", "failed"):
            completed_at = datetime.now(UTC).isoformat()

        try:
            self._admin.table("processing_events").insert(
                {
                    "project_id": project_id,
                    "job_id": job_id,
                    "stage": stage,
                    "status": status,
                    "message": message,
                    "metadata": metadata or {},
                    "completed_at": completed_at,
                }
            ).execute()
        except Exception as exc:
            print(f"[processing-events] failed to emit: {exc}")

File: backend/src/services/processing-events.ts
import type { AppSupabaseClient } from "../lib/supabase.js";
import type { Database } from "../types/database.js";

type EventStatus = Database["public"]["Enums"]["processing_event_status"];

export class ProcessingEventsService {
  constructor(private readonly admin: AppSupabaseClient) {}

  async emit(input: {
    projectId: string;
    jobId?: string;
    stage: string;
    status: EventStatus;
    message?: string;
    metadata?: Record<string, unknown>;
  }) {
    const { error } = await this.admin.from("processing_events").insert({
      project_id: input.projectId,
      job_id: input.jobId ?? null,
      stage: input.stage,
      status: input.status,
      message: input.message ?? null,
      metadata: input.metadata ?? {},
      completed_at: input.status === "completed" || input.status === "failed" ? new Date().toISOString() : null,
    });

    if (error) {
      console.warn("[processing-events] failed to emit:", error.message);
    }
  }
}

export function mapProcessingEvent(row: {
  id: string;
  stage: string;
  status: EventStatus;
  message: string | null;
  started_at: string;
  completed_at: string | null;
}) {
  return {
    id: row.id,
    stage: row.stage,
    status: row.status,
    message: row.message ?? undefined,
    startedAt: row.started_at,
    completedAt: row.completed_at ?? undefined,
  };
}
